### ISSUE TITLE
'll' operations fail when writeback cache is enabled

### ISSUE THREAD

**Issue by Roay** [libfuse/libfuse#323]
Using FUSE3.2.6 to create file system, create a file on the mount directory, the file size is 1G. Format the file and mount it to the directory. After a few seconds, see the directory through the "ll" command.
```
#df -h 
fuse_hello1          1.0G  1.0G     0 100% /root/libfuse-fuse-3.2.6/example/fileDir

#ll -h /root/libfuse-fuse-3.2.6/example/fileDir
total 512K
-rw-r--r-- 1 root root 1.0G Nov 16 15:46 file

#mkfs.xfs /root/libfuse-fuse-3.2.6/example/fileDir/file
#mount /root/libfuse-fuse-3.2.6/example/fileDir/file /root/libfuse-fuse-3.2.6/example/mountDir
```
- When writeback cache is enabled, the error occurs.
```
#ll /root/libfuse-fuse-3.2.6/example/mountDir
ls: cannot access mountDir: Input/output error
```
- When writeback cache is disabled, successful.
```
# ll mountDir/
total 0
```
enable writeback_cache：conn->want |= FUSE_CAP_WRITEBACK_CACHE;
Incorrect usage or other reasons?

**Comment by Nikratio** [comment]
Thanks for the report! What file system are you using? Are you able to reproduce the problem with `examples/passthrough_ll.c`? Most likely the problem is in the filesystem rather than libfuse.

**Comment by Roay** [comment]
linux kernel:4.18
FS:  xfs

**Comment by Roay** [comment]
Sorry, I can't understand the description of "passthrough_ll.c", so I can't reproduce the problem, but you can easily reproduce the problem according to my code. Thank you.
#define FUSE_USE_VERSION 31

#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <sys/socket.h>
#include <arpa/inet.h>
#include <netinet/in.h>
#include <signal.h>
#include <stdlib.h>
#include <errno.h>
#include <stddef.h>
#include <stdint.h>
#include <syslog.h>
#include <pthread.h>
#include <fuse3/fuse.h>
#include <fuse3/fuse_common.h>

#define ROOT_INODE               1
#define FILE_INODE               2

struct fuse_private {
    off_t                        file_size;
    u_int                        blocksize;
    u_int                        num_blocks;
    time_t                       start_time;
    time_t                       file_atime;
    time_t                       file_mtime;
    pthread_mutex_t              mutex;
    char                         *file_name;
    char                         *cache_name;
    int                          cache_fd;
};
struct fuse_private *priv;

/* FUSE functions */
static void *fuse_op_init(struct fuse_conn_info *conn, struct fuse_config *cfg);
static void fuse_op_destroy(void *data);
static int fuse_op_getattr(const char *path, struct stat *st, struct fuse_file_info *fi);
static int fuse_op_readdir(const char *path, void *buf, fuse_fill_dir_t filler,
    off_t offset, struct fuse_file_info *fi, enum fuse_readdir_flags flag);
static int fuse_op_open(const char *path, struct fuse_file_info *fi);
static int fuse_op_release(const char *path, struct fuse_file_info *fi);
static int fuse_op_read(const char *path, char *buf, size_t size, off_t offset,
    struct fuse_file_info *fi);
static int fuse_op_write(const char *path, const char *buf, size_t size, off_t offset,
    struct fuse_file_info *fi);
static int fuse_op_statfs(const char *path, struct statvfs *st);
static int fuse_op_truncate(const char *path, off_t size, struct fuse_file_info *fi);
static int fuse_op_flush(const char *path, struct fuse_file_info *fi);
static int fuse_op_fsync(const char *path, int isdatasync, struct fuse_file_info *fi);
static int test_write(uint64_t fd, const char *filename, off_t offset, const void *data, size_t len);
static int test_read(uint64_t fd, const char *filename, off_t offset, void *data, size_t len);
static void fuse_op_getattr_file(struct stat *st);


/* FUSE operations */
const struct fuse_operations s3backer_fuse_ops = {
    .init       = fuse_op_init,
    .destroy    = fuse_op_destroy,
    .getattr    = fuse_op_getattr,
    .readdir    = fuse_op_readdir,
    .open       = fuse_op_open,
    .read       = fuse_op_read,
    .write      = fuse_op_write,
    .statfs     = fuse_op_statfs,
    .truncate   = fuse_op_truncate,
    .flush      = fuse_op_flush,
    .fsync      = fuse_op_fsync,
    .release    = fuse_op_release,
};

static void *
fuse_op_init(struct fuse_conn_info *conn, struct fuse_config *cfg)
{
	//  LL_ENABLE
//	conn->want |= FUSE_CAP_WRITEBACK_CACHE;

    return NULL;
}

static int
test_write(uint64_t fd, const char *filename, off_t offset, const void *data, size_t len)
{
    size_t sofar;
    ssize_t r;

    for (sofar = 0; sofar < len; sofar += r) {
        const off_t posn = offset + sofar;

        if ((r = pwrite(fd, (const char *)data + sofar, len - sofar, offset + sofar)) == -1) {
        	 printf("error writing cache file '%s' at offset %ju: %s\n",
                    filename, (uintmax_t)posn, strerror(r));
            return r;
        }
    }
    return 0;
}

static int
test_read(uint64_t fd, const char *filename, off_t offset, void *data, size_t len)
{
    size_t sofar;
    ssize_t r;

    for (sofar = 0; sofar < len; sofar += r) {
        const off_t posn = offset + sofar;

        if ((r = pread(fd, (char *)data + sofar, len - sofar, offset + sofar)) == -1) {
        	 printf("error reading cache file '%s' at offset %ju: %s\n",
                    filename, (uintmax_t)posn, strerror(r));
            return r;
        }
        if (r == 0) {
        	 printf("error reading cache file '%s' at offset %ju: file is truncated\n",
                    filename, (uintmax_t)posn);
            return EINVAL;
        }
    }
    return 0;
}

static int
fuse_op_getattr(const char *path, struct stat *st, struct fuse_file_info *fi)
{
    memset(st, 0, sizeof(*st));
    if (strcmp(path, "/") == 0) {
        st->st_mode = S_IFDIR | 0755;
        st->st_nlink = 2;
        st->st_ino = ROOT_INODE;
        st->st_uid = getuid();
        st->st_gid = getgid();
        if (priv != NULL) {
            st->st_atime = priv->start_time;
            st->st_mtime = priv->start_time;
            st->st_ctime = priv->start_time;
        }
        return 0;
    }
    if (priv == NULL)
        return -ENOENT;
    if (*path == '/' && strcmp(path + 1, priv->file_name) == 0) {
        fuse_op_getattr_file(st);
        return 0;
    }
    return -ENOENT;
}

static void
fuse_op_getattr_file(struct stat *st)
{
    st->st_mode = S_IFREG | 0644;
    st->st_nlink = 1;
    st->st_ino = FILE_INODE;
    st->st_uid = getuid();
    st->st_gid = getgid();
    st->st_size = priv->file_size;
    st->st_blksize = priv->blocksize;
    st->st_blocks = priv->num_blocks;
    st->st_atime = priv->file_atime;
    st->st_mtime = priv->file_mtime;
    st->st_ctime = priv->start_time;
}

static int
fuse_op_readdir(const char *path, void *buf, fuse_fill_dir_t filler,
    off_t offset, struct fuse_file_info *fi, enum fuse_readdir_flags flag)
{
    (void)offset;
    (void)fi;
    if (strcmp(path, "/") != 0)
        return -ENOENT;
    if (filler(buf, ".", NULL, 0, 0) != 0)
        return -ENOMEM;
    if (filler(buf, "..", NULL, 0, 0) != 0)
        return -ENOMEM;
    if (priv != NULL) {
        if (filler(buf, priv->file_name, NULL, 0, 0) != 0)
            return -ENOMEM;
    }
    return 0;
}

static int
fuse_op_open(const char *path, struct fuse_file_info *fi)
{
    if (*path == '/' && strcmp(path + 1, priv->file_name) == 0) {
        fi->fh = 0;
        priv->file_atime = time(NULL);
//		fi->direct_io = 1;
        return 0;
    }

    return -ENOENT;
}

static int
fuse_op_read(const char *path, char *buf, size_t size, off_t offset,
    struct fuse_file_info *fi)
{
    size_t orig_size = size;

    /* Check for end of file */
    if (offset > priv->file_size) {
    	printf("read offset=0x%jx size=0x%jx out of range\n", (uintmax_t)offset, (uintmax_t)size);
        return -ESPIPE;
    }
    if (offset + size > priv->file_size) {
        size = priv->file_size - offset;
        orig_size = size;
    }

    if (size == 0)
        return 0;
    pthread_mutex_lock(&priv->mutex);
    printf("fuse_read1>>>>>>>>>>>>>>> offset=0x%jx size=0x%jx\n",(uintmax_t)offset, (uintmax_t)size);
    test_read(priv->cache_fd, priv->cache_name, offset, buf, size);
    printf("fuse_read2>>>>>>>>>>>>>>> offset=0x%jx size=0x%jx\n",(uintmax_t)offset, (uintmax_t)size);
    pthread_mutex_unlock(&priv->mutex);
    return orig_size;
}

static int fuse_op_write(const char *path, const char *buf, size_t size,
    off_t offset, struct fuse_file_info *fi)
{
    size_t orig_size = size;

    /* Check for end of file */
    if (offset > priv->file_size) {
    	printf("write offset=0x%jx size=0x%jx out of range\n", (uintmax_t)offset, (uintmax_t)size);
        return -ESPIPE;
    }
    if (offset + size > priv->file_size) {
        size = priv->file_size - offset;
        orig_size = size;
    }

    if (size == 0)
        return 0;
    pthread_mutex_lock(&priv->mutex);
    printf("fuse_write1>>>>>>>>>>>>>>> offset=0x%jx size=0x%jx\n", (uintmax_t)offset, (uintmax_t)size);
    test_write(priv->cache_fd, priv->cache_name, offset, buf, size);
    printf("fuse_write2>>>>>>>>>>>>>>> offset=0x%jx size=0x%jx\n",(uintmax_t)offset, (uintmax_t)size);
    pthread_mutex_unlock(&priv->mutex);
    return orig_size;
}


static int
fuse_op_statfs(const char *path, struct statvfs *st)
{
    st->f_bsize = priv->blocksize;
    st->f_frsize = priv->blocksize;
    st->f_blocks = priv->num_blocks;
    st->f_bfree = 0;
    st->f_bavail = 0;
    st->f_files = 3;
    st->f_ffree = 0;
    st->f_favail = 0;
    return 0;
}

static int
fuse_op_truncate(const char *path, off_t size, struct fuse_file_info *fi)
{
    return 0;
}

static int
fuse_op_release(const char *path, struct fuse_file_info *fi)
{
    return 0;
}

static int
fuse_op_flush(const char *path, struct fuse_file_info *fi)
{
    return 0;
}

static int
fuse_op_fsync(const char *path, int isdatasync, struct fuse_file_info *fi)
{
    return 0;
}

static void
fuse_op_destroy(void *data)
{
    if (priv == NULL)
        return;

    close(priv->cache_fd);
    pthread_mutex_destroy(&priv->mutex);

    free(priv);
}
int main(int argc, char *argv[])
{
	int r;

    if ((priv = calloc(1, sizeof(*priv))) == NULL) {
        printf("fuse_op_init(): %s\n", strerror(errno));
        exit(1);
    }
    priv->file_name = "file";
    priv->start_time = time(NULL);
    priv->file_atime = priv->start_time;
    priv->file_mtime = priv->start_time;
    priv->blocksize = 1048576;
    priv->num_blocks = 1024;
    priv->file_size = priv->blocksize * priv->num_blocks;

    //test
    priv->cache_name = "/usr/src/workspace/test/src/testCache";
    if ((priv->cache_fd = open(priv->cache_name, O_RDWR|O_CREAT|O_EXCL, 0644)) == -1) {
        r = errno;
        printf("can't create file '%s': %s", priv->cache_name, strerror(r));
        return r;
    }

    return fuse_main(argc, argv, &s3backer_fuse_ops, NULL);
}

**Comment by Nikratio** [comment]
I'm sorry, but I do not have enough time to review this. Unless you have some clear evidence that the problem is in libfuse rather then your code, I can't justify spending time on investigating further.

**Comment by Nikratio** [comment]
Could you please provide *detailed*, *step-by-step* instructions for reproducing this problem?

On my system:

```
$ example/passthrough_ll ~/tmp/mnt &
[1] 12963

$ dd if=/dev/zero of=bigfile bs=1k seek=1024k count=1
1+0 records in
1+0 records out
1024 bytes (1.0 kB, 1.0 KiB) copied, 0.000100174 s, 10.2 MB/s

$ ls -lh bigfile 
-rw-r--r-- 1 nikratio nikratio 1.1G Nov 19 12:36 bigfile

$ mkfs.xfs bigfile 
meta-data=bigfile                isize=512    agcount=4, agsize=65536 blks
         =                       sectsz=512   attr=2, projid32bit=1
         =                       crc=1        finobt=1, sparse=0, rmapbt=0, reflink=0
data     =                       bsize=4096   blocks=262144, imaxpct=25
         =                       sunit=0      swidth=0 blks
naming   =version 2              bsize=4096   ascii-ci=0 ftype=1
log      =internal log           bsize=4096   blocks=2560, version=2
         =                       sectsz=512   sunit=0 blks, lazy-count=1
realtime =none                   extsz=4096   blocks=0, rtextents=0
$
```

**Comment by Roay** [comment]
```
# gcc -Wall passthrough_ll.c `pkg-config fuse3 --cflags --libs` -o passthrough_ll
# 
# ll -h passthrough_ll
-rwxr-xr-x 1 root root 25K Nov 19 20:46 passthrough_ll
# ./passthrough_ll fileDir/
# cd fileDir/

# dd if=/dev/zero of=bigfile bs=1k seek=1024k count=1
dd: failed to truncate to 1073741824 bytes in output file ‘bigfile’: Function not implemented

# dd if=/dev/zero of=bigfile bs=1M count=1024
1024+0 records in
1024+0 records out
1073741824 bytes (1.1 GB) copied, 11.9256 s, 90.0 MB/s

# mkfs.xfs bigfile 
meta-data=bigfile                isize=512    agcount=4, agsize=65536 blks
         =                       sectsz=512   attr=2, projid32bit=1
         =                       crc=1        finobt=0, sparse=0
data     =                       bsize=4096   blocks=262144, imaxpct=25
         =                       sunit=0      swidth=0 blks
naming   =version 2              bsize=4096   ascii-ci=0 ftype=1
log      =internal log           bsize=4096   blocks=2560, version=2
         =                       sectsz=512   sunit=0 blks, lazy-count=1
realtime =none                   extsz=4096   blocks=0, rtextents=0
mkfs.xfs: pwrite64 failed: Invalid argument
```

**Comment by Roay** [comment]
But in other directory (non mounted directory), like "/mnt".
```
# pwd
/mnt

# dd if=/dev/zero of=bigfile bs=1k seek=1k seek=1024k count=1
1+0 records in
1+0 records out
1024 bytes (1.0 kB) copied, 0.000477441 s, 2.1 MB/s

# ll -h
total 4.0K
-rw-r--r-- 1 root root 1.1G Nov 19 20:59 bigfile
```

**Comment by Nikratio** [comment]
Please update to libfuse 3.3 and try again.

**Comment by Roay** [comment]
Sorry, under the fuse3.3 version, `dd if=/dev/zero of=bigfile bs=1k seek=1k seek=1024k count=1` executed successfully, but mkfs.xfs still reported errors.
```
# pkg-config --modversion fuse
3.3.0
# dd if=/dev/zero of=bigfile bs=1k seek=1024k count=1
1+0 records in
1+0 records out
1024 bytes (1.0 kB) copied, 0.00049762 s, 2.1 MB/s
# ll bigfile -h
-rw-r--r-- 1 root root 1.1G Nov 20 09:19 bigfile
# mkfs.xfs bigfile 
meta-data=bigfile                isize=512    agcount=4, agsize=65536 blks
         =                       sectsz=512   attr=2, projid32bit=1
         =                       crc=1        finobt=0, sparse=0
data     =                       bsize=4096   blocks=262144, imaxpct=25
         =                       sunit=0      swidth=0 blks
naming   =version 2              bsize=4096   ascii-ci=0 ftype=1
log      =internal log           bsize=4096   blocks=2560, version=2
         =                       sectsz=512   sunit=0 blks, lazy-count=1
realtime =none                   extsz=4096   blocks=0, rtextents=0
mkfs.xfs: pwrite64 failed: Invalid argument
```

- The fuse debug info(`./passthrough_ll testDir/ -d`):
unique: 572, opcode: LOOKUP (1), nodeid: 1, insize: 48, pid: 27241
lo_lookup(parent=1, name=bigfile)
  1/bigfile -> 140583473656976
   unique: 572, success, outsize: 144
unique: 573, opcode: LOOKUP (1), nodeid: 1, insize: 48, pid: 27512
lo_lookup(parent=1, name=bigfile)
  1/bigfile -> 140583473656976
   unique: 573, success, outsize: 144
unique: 574, opcode: OPEN (14), nodeid: 140583473656976, insize: 48, pid: 27512
lo_open(ino=140583473656976, flags=32768)
   unique: 574, success, outsize: 32
unique: 575, opcode: IOCTL (39), nodeid: 140583473656976, insize: 72, pid: 27512
   unique: 575, error: -38 (Function not implemented), outsize: 16
unique: 576, opcode: FLUSH (25), nodeid: 140583473656976, insize: 64, pid: 27512
   unique: 576, success, outsize: 16
unique: 577, opcode: RELEASE (18), nodeid: 140583473656976, insize: 64, pid: 0
   unique: 577, success, outsize: 16
unique: 578, opcode: OPEN (14), nodeid: 140583473656976, insize: 48, pid: 27512
lo_open(ino=140583473656976, flags=32768)
   unique: 578, success, outsize: 32
unique: 579, opcode: OPEN (14), nodeid: 140583473656976, insize: 48, pid: 27512
lo_open(ino=140583473656976, flags=49154)
   unique: 579, success, outsize: 32
unique: 580, opcode: IOCTL (39), nodeid: 140583473656976, insize: 72, pid: 27512
   unique: 580, error: -38 (Function not implemented), outsize: 16
unique: 581, opcode: FLUSH (25), nodeid: 140583473656976, insize: 64, pid: 27512
   unique: 581, success, outsize: 16
unique: 582, opcode: RELEASE (18), nodeid: 140583473656976, insize: 64, pid: 0
unique: 583, opcode: OPEN (14), nodeid: 140583473656976, insize: 48, pid: 27512
lo_open(ino=140583473656976, flags=32768)
   unique: 582, success, outsize: 16
   unique: 583, success, outsize: 32
unique: 584, opcode: IOCTL (39), nodeid: 140583473656976, insize: 72, pid: 27512
   unique: 584, error: -38 (Function not implemented), outsize: 16
unique: 585, opcode: FLUSH (25), nodeid: 140583473656976, insize: 64, pid: 27512
   unique: 585, success, outsize: 16
unique: 586, opcode: RELEASE (18), nodeid: 140583473656976, insize: 64, pid: 0
   unique: 586, success, outsize: 16
unique: 587, opcode: OPEN (14), nodeid: 140583473656976, insize: 48, pid: 27512
lo_open(ino=140583473656976, flags=32768)
   unique: 587, success, outsize: 32
unique: 588, opcode: READ (15), nodeid: 140583473656976, insize: 80, pid: 27512
lo_read(ino=140583473656976, size=4096, off=1073676288)
   unique: 588, success, outsize: 4112
unique: 589, opcode: GETATTR (3), nodeid: 140583473656976, insize: 56, pid: 27512
   unique: 589, success, outsize: 120
unique: 590, opcode: READ (15), nodeid: 140583473656976, insize: 80, pid: 27512
lo_read(ino=140583473656976, size=4096, off=1073733632)
   unique: 590, success, outsize: 4112
unique: 591, opcode: GETATTR (3), nodeid: 140583473656976, insize: 56, pid: 27512
   unique: 591, success, outsize: 120
unique: 592, opcode: READ (15), nodeid: 140583473656976, insize: 80, pid: 27512
lo_read(ino=140583473656976, size=4096, off=0)
   unique: 592, success, outsize: 4112
unique: 593, opcode: GETATTR (3), nodeid: 140583473656976, insize: 56, pid: 27512
   unique: 593, success, outsize: 120
unique: 594, opcode: READ (15), nodeid: 140583473656976, insize: 80, pid: 27512
lo_read(ino=140583473656976, size=4096, off=4096)
   unique: 594, success, outsize: 4112
unique: 595, opcode: GETATTR (3), nodeid: 140583473656976, insize: 56, pid: 27512
   unique: 595, success, outsize: 120
unique: 596, opcode: READ (15), nodeid: 140583473656976, insize: 80, pid: 27512
lo_read(ino=140583473656976, size=4096, off=1073741824)
   unique: 596, success, outsize: 1040
unique: 597, opcode: GETATTR (3), nodeid: 140583473656976, insize: 56, pid: 27512
   unique: 597, success, outsize: 120
unique: 598, opcode: READ (15), nodeid: 140583473656976, insize: 80, pid: 27512
lo_read(ino=140583473656976, size=4096, off=1073610752)
   unique: 598, success, outsize: 4112
unique: 599, opcode: GETATTR (3), nodeid: 140583473656976, insize: 56, pid: 27512
   unique: 599, success, outsize: 120
unique: 600, opcode: READ (15), nodeid: 140583473656976, insize: 80, pid: 27512
lo_read(ino=140583473656976, size=4096, off=1073709056)
   unique: 600, success, outsize: 4112
unique: 601, opcode: GETATTR (3), nodeid: 140583473656976, insize: 56, pid: 27512
   unique: 601, success, outsize: 120
unique: 602, opcode: READ (15), nodeid: 140583473656976, insize: 80, pid: 27512
lo_read(ino=140583473656976, size=4096, off=1073537024)
   unique: 602, success, outsize: 4112
unique: 603, opcode: GETATTR (3), nodeid: 140583473656976, insize: 56, pid: 27512
   unique: 603, success, outsize: 120
unique: 604, opcode: READ (15), nodeid: 140583473656976, insize: 80, pid: 27512
lo_read(ino=140583473656976, size=4096, off=1073438720)
   unique: 604, success, outsize: 4112
unique: 605, opcode: GETATTR (3), nodeid: 140583473656976, insize: 56, pid: 27512
   unique: 605, success, outsize: 120
unique: 606, opcode: READ (15), nodeid: 140583473656976, insize: 80, pid: 27512
lo_read(ino=140583473656976, size=4096, off=1073393664)
   unique: 606, success, outsize: 4112
unique: 607, opcode: GETATTR (3), nodeid: 140583473656976, insize: 56, pid: 27512
   unique: 607, success, outsize: 120
unique: 608, opcode: READ (15), nodeid: 140583473656976, insize: 80, pid: 27512
lo_read(ino=140583473656976, size=4096, off=1073364992)
   unique: 608, success, outsize: 4112
unique: 609, opcode: GETATTR (3), nodeid: 140583473656976, insize: 56, pid: 27512
   unique: 609, success, outsize: 120
unique: 610, opcode: READ (15), nodeid: 140583473656976, insize: 80, pid: 27512
lo_read(ino=140583473656976, size=4096, off=1073274880)
   unique: 610, success, outsize: 4112
unique: 611, opcode: GETATTR (3), nodeid: 140583473656976, insize: 56, pid: 27512
   unique: 611, success, outsize: 120
unique: 612, opcode: READ (15), nodeid: 140583473656976, insize: 80, pid: 27512
lo_read(ino=140583473656976, size=4096, off=1073242112)
   unique: 612, success, outsize: 4112
unique: 613, opcode: GETATTR (3), nodeid: 140583473656976, insize: 56, pid: 27512
   unique: 613, success, outsize: 120
unique: 614, opcode: READ (15), nodeid: 140583473656976, insize: 80, pid: 27512
lo_read(ino=140583473656976, size=4096, off=1073233920)
   unique: 614, success, outsize: 4112
unique: 615, opcode: GETATTR (3), nodeid: 140583473656976, insize: 56, pid: 27512
   unique: 615, success, outsize: 120
unique: 616, opcode: READ (15), nodeid: 140583473656976, insize: 80, pid: 27512
lo_read(ino=140583473656976, size=4096, off=1073254400)
   unique: 616, success, outsize: 4112
unique: 617, opcode: GETATTR (3), nodeid: 140583473656976, insize: 56, pid: 27512
   unique: 617, success, outsize: 120
unique: 618, opcode: READ (15), nodeid: 140583473656976, insize: 80, pid: 27512
lo_read(ino=140583473656976, size=4096, off=1072160768)
   unique: 618, success, outsize: 4112
unique: 619, opcode: GETATTR (3), nodeid: 140583473656976, insize: 56, pid: 27512
   unique: 619, success, outsize: 120
unique: 620, opcode: READ (15), nodeid: 140583473656976, insize: 80, pid: 27512
lo_read(ino=140583473656976, size=4096, off=1073737728)
   unique: 620, success, outsize: 4112
unique: 621, opcode: GETATTR (3), nodeid: 140583473656976, insize: 56, pid: 27512
   unique: 621, success, outsize: 120
unique: 622, opcode: READ (15), nodeid: 140583473656976, insize: 80, pid: 27512
lo_read(ino=140583473656976, size=4096, off=1048576)
   unique: 622, success, outsize: 4112
unique: 623, opcode: GETATTR (3), nodeid: 140583473656976, insize: 56, pid: 27512
   unique: 623, success, outsize: 120
unique: 624, opcode: READ (15), nodeid: 140583473656976, insize: 80, pid: 27512
lo_read(ino=140583473656976, size=4096, off=12288)
   unique: 624, success, outsize: 4112
unique: 625, opcode: GETATTR (3), nodeid: 140583473656976, insize: 56, pid: 27512
   unique: 625, success, outsize: 120
unique: 626, opcode: READ (15), nodeid: 140583473656976, insize: 80, pid: 27512
lo_read(ino=140583473656976, size=4096, off=28672)
   unique: 626, success, outsize: 4112
unique: 627, opcode: GETATTR (3), nodeid: 140583473656976, insize: 56, pid: 27512
   unique: 627, success, outsize: 120
unique: 628, opcode: READ (15), nodeid: 140583473656976, insize: 80, pid: 27512
lo_read(ino=140583473656976, size=4096, off=61440)
   unique: 628, success, outsize: 4112
unique: 629, opcode: GETATTR (3), nodeid: 140583473656976, insize: 56, pid: 27512
   unique: 629, success, outsize: 120
unique: 631, opcode: READ (15), nodeid: 140583473656976, insize: 80, pid: 27512
lo_read(ino=140583473656976, size=12288, off=16384)
unique: 632, opcode: READ (15), nodeid: 140583473656976, insize: 80, pid: 27512
unique: 630, opcode: READ (15), nodeid: 140583473656976, insize: 80, pid: 27512
lo_read(ino=140583473656976, size=4096, off=8192)
   unique: 630, success, outsize: 4112
lo_read(ino=140583473656976, size=28672, off=32768)
unique: 633, opcode: READ (15), nodeid: 140583473656976, insize: 80, pid: 27512
lo_read(ino=140583473656976, size=73728, off=65536)
   unique: 632, success, outsize: 28688
   unique: 631, success, outsize: 12304
   unique: 633, success, outsize: 73744
unique: 634, opcode: READ (15), nodeid: 140583473656976, insize: 80, pid: 27512
lo_read(ino=140583473656976, size=122880, off=139264)
   unique: 634, success, outsize: 122896
unique: 635, opcode: GETATTR (3), nodeid: 140583473656976, insize: 56, pid: 27512
   unique: 635, success, outsize: 120
unique: 636, opcode: READ (15), nodeid: 140583473656976, insize: 80, pid: 27512
lo_read(ino=140583473656976, size=131072, off=262144)
   unique: 636, success, outsize: 131088
unique: 637, opcode: READ (15), nodeid: 140583473656976, insize: 80, pid: 27512
lo_read(ino=140583473656976, size=131072, off=393216)
   unique: 637, success, outsize: 131088
unique: 638, opcode: GETATTR (3), nodeid: 140583473656976, insize: 56, pid: 27512
   unique: 638, success, outsize: 120
unique: 639, opcode: READ (15), nodeid: 140583473656976, insize: 80, pid: 27512
unique: 640, opcode: READ (15), nodeid: 140583473656976, insize: 80, pid: 27512
unique: 641, opcode: READ (15), nodeid: 140583473656976, insize: 80, pid: 27512
lo_read(ino=140583473656976, size=16384, off=1073217536)
lo_read(ino=140583473656976, size=4096, off=1073238016)
   unique: 640, success, outsize: 4112
lo_read(ino=140583473656976, size=8192, off=1073246208)
unique: 642, opcode: READ (15), nodeid: 140583473656976, insize: 80, pid: 27512
   unique: 641, success, outsize: 8208
lo_read(ino=140583473656976, size=16384, off=1073258496)
   unique: 639, success, outsize: 16400
   unique: 642, success, outsize: 16400
unique: 643, opcode: READ (15), nodeid: 140583473656976, insize: 80, pid: 27512
lo_read(ino=140583473656976, size=69632, off=1073278976)
   unique: 643, success, outsize: 69648
unique: 644, opcode: READ (15), nodeid: 140583473656976, insize: 80, pid: 27512
lo_read(ino=140583473656976, size=16384, off=1073348608)
unique: 645, opcode: READ (15), nodeid: 140583473656976, insize: 80, pid: 27512
lo_read(ino=140583473656976, size=24576, off=1073369088)
unique: 646, opcode: READ (15), nodeid: 140583473656976, insize: 80, pid: 27512
lo_read(ino=140583473656976, size=40960, off=1073397760)
unique: 647, opcode: READ (15), nodeid: 140583473656976, insize: 80, pid: 27512
   unique: 646, success, outsize: 40976
   unique: 645, success, outsize: 24592
lo_read(ino=140583473656976, size=36864, off=1073442816)
   unique: 644, success, outsize: 16400
   unique: 647, success, outsize: 36880
unique: 648, opcode: GETATTR (3), nodeid: 140583473656976, insize: 56, pid: 27512
   unique: 648, success, outsize: 120
unique: 649, opcode: READ (15), nodeid: 140583473656976, insize: 80, pid: 27512
lo_read(ino=140583473656976, size=57344, off=1073479680)
   unique: 649, success, outsize: 57360
unique: 650, opcode: READ (15), nodeid: 140583473656976, insize: 80, pid: 27512
lo_read(ino=140583473656976, size=69632, off=1073541120)
   unique: 650, success, outsize: 69648
unique: 651, opcode: READ (15), nodeid: 140583473656976, insize: 80, pid: 27512
lo_read(ino=140583473656976, size=61440, off=1073614848)
unique: 652, opcode: READ (15), nodeid: 140583473656976, insize: 80, pid: 27512
lo_read(ino=140583473656976, size=28672, off=1073680384)
   unique: 651, success, outsize: 61456
   unique: 652, success, outsize: 28688
unique: 653, opcode: READ (15), nodeid: 140583473656976, insize: 80, pid: 27512
lo_read(ino=140583473656976, size=20480, off=1073713152)
   unique: 653, success, outsize: 20496
unique: 654, opcode: GETATTR (3), nodeid: 140583473656976, insize: 56, pid: 27512
   unique: 654, success, outsize: 120
unique: 655, opcode: READ (15), nodeid: 140583473656976, insize: 80, pid: 27512
lo_read(ino=140583473656976, size=4096, off=2097152)
   unique: 655, success, outsize: 4112
unique: 656, opcode: FLUSH (25), nodeid: 140583473656976, insize: 64, pid: 27512
   unique: 656, success, outsize: 16
unique: 657, opcode: RELEASE (18), nodeid: 140583473656976, insize: 64, pid: 0
   unique: 657, success, outsize: 16
**unique: 658, opcode: IOCTL (39), nodeid: 140583473656976, insize: 72, pid: 27512
   unique: 658, error: -38 (Function not implemented), outsize: 16**
unique: 659, opcode: WRITE (16), nodeid: 140583473656976, insize: 128080, pid: 27512
lo_write(ino=140583473656976, size=128000, off=0)
unique: 660, opcode: WRITE (16), nodeid: 140583473656976, insize: 3152, pid: 27512
lo_write(ino=140583473656976, size=3072, off=128000)
   unique: 659, success, outsize: 24
   **unique: 660, error: -22 (Invalid argument), outsize: 16**
unique: 661, opcode: FLUSH (25), nodeid: 140583473656976, insize: 64, pid: 27512
   unique: 661, success, outsize: 16
unique: 662, opcode: RELEASE (18), nodeid: 140583473656976, insize: 64, pid: 0
   unique: 662, success, outsize: 16

**Comment by Nikratio** [comment]
Could you please provide detailed, step-by-step instructions for reproducing this problem?

**Comment by Roay** [comment]
```
# gcc -Wall passthrough_ll.c `pkg-config fuse3 --cflags --libs` -o passthrough_ll
# ll -h passthrough_ll
-rwxr-xr-x 1 root root 36K Nov 20 19:57 passthrough_ll
# ./passthrough_ll testDir/
# cd testDir/
# dd if=/dev/zero of=bigfile bs=1k seek=1024k count=1
1+0 records in
1+0 records out
1024 bytes (1.0 kB) copied, 0.000579427 s, 1.8 MB/s
# ll -h bigfile 
-rw-r--r-- 1 root root 1.1G Nov 20 20:00 bigfile
# mkfs.xfs bigfile 
meta-data=bigfile                isize=512    agcount=4, agsize=65536 blks
         =                       sectsz=512   attr=2, projid32bit=1
         =                       crc=1        finobt=0, sparse=0
data     =                       bsize=4096   blocks=262144, imaxpct=25
         =                       sunit=0      swidth=0 blks
naming   =version 2              bsize=4096   ascii-ci=0 ftype=1
log      =internal log           bsize=4096   blocks=2560, version=2
         =                       sectsz=512   sunit=0 blks, lazy-count=1
realtime =none                   extsz=4096   blocks=0, rtextents=0
mkfs.xfs: pwrite64 failed: Invalid argument
```

**Comment by Roay** [comment]
# uname -r
4.18.15-1.el7.elrepo.x86_64
# pkg-config --modVersion fuse
Unknown option --modVersion
[root@hadoop testDir]# uname -r
4.18.15-1.el7.elrepo.x86_64
# pkg-config --modversion fuse
3.3.0
# python -V
Python 2.7.5
# gcc --version
gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-28)

**Comment by Roay** [comment]
Is this the bug of libfuse?

**Comment by Roay** [comment]
I tested the same problem in a new environment in centos7.4 and in ubuntu16.04

**Comment by Roay** [comment]
#  mkfs.xfs -V
mkfs.xfs version 4.5.0

**Comment by Roay** [comment]
What is your operating system and what is the kernel version?
In CentOS 7, the latest version only supports `mkfs.xfs version 4.5.0`

**Comment by Roay** [comment]
Thanks, the problem was solved after updating xfsprogs to v4.9.0.

I did another test. I added `conn->want |= FUSE_CAP_WRITEBACK_CACHE;` to the passthrough.c `xmp_init` method. Then did the following
```
# gcc -Wall passthrough.c `pkg-config fuse3 --cflags --libs` -o passthrough
# ll -h passthrough
-rwxr-xr-x 1 root root 19K Nov 24 16:20 passthrough
# ./passthrough testDir/
# cd testDir/
# dd if=/dev/zero of=bigfile bs=1k seek=1024k count=1
1+0 records in
1+0 records out
1024 bytes (1.0 kB) copied, 0.000579427 s, 1.8 MB/s
# ll -h bigfile 
-rw-r--r-- 1 root root 1.1G Nov 24 16:31 bigfile
# mkfs.xfs bigfile 
# mount bigfile /mnt
# ll /mnt/
total 4
drwxr-xr-x  2 root root    6 Nov 24 17:08 ./
drwxr-xr-x 23 root root 4096 Nov 24 16:36 ../
```
**A few minutes later**
```
# ll /mnt/
ls: cannot access '/mnt/': Input/output error
```
**Can it be implemented `writeback cache` through the high-level API ?**
